pkg/brbct/brbct.go

```diff
@@ -0,0 +1,341 @@
+	Crypto              t.ModuleID
+	Hasher              t.ModuleID
+	MerkleProofVerifier t.ModuleID
+}
+
+// DefaultModuleConfig returns a valid module config with default names for all modules.
+func DefaultModuleConfig(consumer t.ModuleID) *ModuleConfig {
+	return &ModuleConfig{
+		Self:                "brbct",
+		Consumer:            consumer,
+		Net:                 "net",
+		Crypto:              "crypto",
+		Hasher:              "hasher",
+		MerkleProofVerifier: "merkleverify",
+	}
+}
+
+// ModuleParams sets the values for the parameters of an instance of the protocol.
+// All replicas are expected to use identical module parameters.
+type ModuleParams struct {
+	InstanceUID []byte     // unique identifier for this instance of BRB, used to prevent cross-instance replay attacks
+	AllNodes    []t.NodeID // the list of participating nodes
+	Leader      t.NodeID   // the id of the leader of the instance
+}
+
+// GetN returns the total number of nodes.
+func (params *ModuleParams) GetN() int {
+	return len(params.AllNodes)
+}
+
+// GetF returns the maximum tolerated number of faulty nodes.
+func (params *ModuleParams) GetF() int {
+	return (params.GetN() - 1) / 3
+}
+
+type Accumulator = struct {
+	value string
+	count int
+}
+
+// brbHashModuleState represents the state of the brb module.
+type brbHashModuleState struct {
+	// this variable is not part of the original protocol description, but it greatly simplifies the code
+	proposed []byte
+
+	sentEcho  bool
+	sentReady bool
+	delivered bool
+
+	echos                    map[string][][]byte
+	echoMessagesCount        map[string]int
+	echoMessagesAccumulator  Accumulator
+	readyMessagesReceived    map[string][]bool
+	readyMessagesCount       map[string]int
+	readyMessagesAccumulator Accumulator
+}
+
+func NewModule(mc *ModuleConfig, params *ModuleParams, nodeID t.NodeID) (modules.PassiveModule, error) {
+	m := dsl.NewModule(mc.Self)
+
+	encoder, err := rs.New(params.GetN()-2*params.GetF(), 2*params.GetF())
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to create coder")
+	}
+
+	// upon event <brb, Init> do
+	state := make(map[int64]*brbHashModuleState)
+	lastId := int64(0)
+
+	// upon event <brb, Broadcast | m> do
+	brbpbdsl.UponBroadcastRequest(m, func(id int64, hdata []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		if nodeID != params.Leader {
+			return fmt.Errorf("only the leader node can receive requests")
+		}
+
+		data := []byte{0, 0, 0, 0}
+		binary.LittleEndian.PutUint32(data, uint32(len(hdata)))
+		data = append(data, hdata...)
+
+		chunkSize := int(math.Ceil(float64(len(data)) / float64(params.GetN()-2*params.GetF())))
+		encoded := make([][]byte, params.GetN())
+		for i := range encoded {
+			encoded[i] = make([]byte, chunkSize)
+		}
+
+		for i, in := range encoded[:params.GetN()-2*params.GetF()] {
+			for j := range in {
+				if i*chunkSize+j < len(data) {
+					in[j] = data[i*chunkSize+j]
+				} else {
+					in[j] = 0
+				}
+			}
+		}
```

Comment:
```suggestion
		data := make([]byte, mathutil.Pad(4+len(hdata), nDataShards))
		binary.LittleEndian.PutUint32(data, uint32(len(hdata)))
		copy(data[4:], hdata)

		shardSize := len(data) / nDataShards
		encoded := make([][]byte, params.GetN())

		for i := 0; i < nDataShards; i++ {
			encoded[i] = data[i*shardSize : (i+1)*shardSize]
		}

		// allocate space for the parity shards
		for i := 0; i < nParityShards; i++ {
			encoded[nDataShards+i] = make([]byte, shardSize)
		}
```

This code assumes variables `nDataShards = N-2F` and `nParityShard = 2F` as well as a function `mathutil.Pad` (in package `pkg/util/mathutil`) that would pad the length to be a multiple of `nDataShards`.

Note that the assignment `encoded[i] = data[i*shardSize : (i+1)*shardSize]` does not copy the data.